download/oci_download.go

```diff
@@ -340,40 +344,42 @@ func (d *OCIDownloader) getResolverHost(client *http.Client, urlInfo *url.URL) d
 		Capabilities: docker.HostCapabilityPull | docker.HostCapabilityResolve | docker.HostCapabilityPush,
 		Client:       client,
 		Path:         "/v2",
-		Authorizer:   auth,
+		Authorizer:   &authorizer,
 	}
 
-	return func(string) ([]docker.RegistryHost, error) {
-		return []docker.RegistryHost{registryHost}, nil
+	opts := docker.ResolverOptions{
+		Hosts: func(string) ([]docker.RegistryHost, error) {
+			return []docker.RegistryHost{registryHost}, nil
+		},
 	}
+
+	return docker.NewResolver(opts), nil
 }
 
-func (d *OCIDownloader) getHTTPClient(authHeader *http.Header) (*http.Client, error) {
-	var client *http.Client
-	var err error
-	clientConfig := d.client.Config()
-	if clientConfig != nil && clientConfig.Credentials.ClientTLS != nil {
-		client, err = clientConfig.Credentials.ClientTLS.NewClient(*clientConfig)
-		if err != nil {
-			return nil, fmt.Errorf("can not create a new client: %w", err)
-		}
-	} else {
-		if clientConfig != nil && clientConfig.Credentials.Bearer != nil {
-			client, err = clientConfig.Credentials.Bearer.NewClient(*clientConfig)
-			if err != nil {
-				return nil, fmt.Errorf("can not create a new bearer client: %w", err)
-			}
+type pluginAuthorizer struct {
+	authorizer docker.Authorizer
+	plugin     rest.HTTPAuthPlugin
 
-			authHeader.Add("Authorization",
-				fmt.Sprintf("%s %s",
-					clientConfig.Credentials.Bearer.Scheme,
-					base64.StdEncoding.EncodeToString([]byte(clientConfig.Credentials.Bearer.Token))),
-			)
-		} else {
-			client = &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: clientConfig.AllowInsecureTLS}}}
-		}
+	logger logging.Logger
+}
+
+var _ docker.Authorizer = &pluginAuthorizer{}
+
+func (a *pluginAuthorizer) AddResponses(ctx context.Context, resps []*http.Response) error {
+	return a.authorizer.AddResponses(ctx, resps)
+}
+
+// Authorize uses a rest.HTTPAuthPlugin to Prepare a request before passing it on to the
+// docker.Authorizer
+func (a *pluginAuthorizer) Authorize(ctx context.Context, req *http.Request) error {
+	if err := a.plugin.Prepare(req); err != nil {
+		// Make sure to log this before passing the error back to docker
+		a.logger.Error("Preparing docker request: %v.", err)
+
+		return fmt.Errorf("preparing docker request: %w", err)
 	}
-	return client, nil
+
+	return a.authorizer.Authorize(ctx, req)
```

Comment:
The normal flow of operation (if an auth plugin is used) is that the `a.plugin`'s `rest.HTTPAuthPlugin.Prepare` should set the Authorization header while the `a.authorizer`'s `docker.Authorizer.Authorize` will do a noop. Initially, the `docker.Authorizer` has no auth handlers defined and simply skips authorization ([code](https://github.com/DerGut/opa/blob/7928ef014e405c5538104a7cde151a7375661599/vendor/github.com/containerd/containerd/remotes/docker/authorizer.go#L120-L124)).

If authentication fails and the registry replies with a `401 Unauthorized` status, the `docker.Authorizer` will call `Authorizer.AddResponses()` to add such auth handlers ([code](https://github.com/DerGut/opa/blob/7928ef014e405c5538104a7cde151a7375661599/vendor/github.com/containerd/containerd/remotes/docker/resolver.go#L606)).

The handlers implementations look at the`WWW-Authenticate` header of the response and try to handle the according challenges by resending the request with Basic or Bearer auth tokens based on credentials passed to `NewDockerAuthorizer()`.

On closer look, this doesn't do much because we don't pass any credentials to `NewDockerAuthorizer()` but instead rely solely on the `HTTPAuthPlugin` to handle Basic and Bearer auth. Instead of sending unauthenticated requests and only adding an Authorization header on the second try, we add the header preemptively for every request.

I mostly left the call to `docker.Authorizer.Authorize()` to be "safe" and not change any behavior that I was unaware of. Now, I believe that we can safely remove it.

Reply:
Addressed in https://github.com/open-policy-agent/opa/pull/5560/commits/c5428d9f4df2032d879c2f0156555b7ec622c6ba

Reply:
Thanks for the detailed explanation.